esmvalcore/preprocessor/_multimodel.py

```diff
@@ -372,22 +349,89 @@ def multi_model_statistics(products, span, statistics, output_products=None):
     for statistic in statistics:
-        # Compute statistic
         statistic_cube = _assemble_data(cubes, statistic, span)
+        statistics_cubes[statistic] = statistic_cube
 
-        if output_products:
-            # Add to output product and log provenance
-            statistic_product = output_products[statistic]
-            statistic_product.cubes = [statistic_cube]
-            for product in products:
-                statistic_product.wasderivedfrom(product)
-            logger.info("Generated %s", statistic_product)
-            statistic_products.add(statistic_product)
-        else:
-            statistic_products[statistic] = statistic_cube
+    return statistics_cubes
+
+
+def _multiproduct_statistics(products,
+                             statistics,
+                             output_products,
+                             span='overlap'):
+    """Compute statistics across ESMValCore products.
+
+    Extract cubes from products, calculate the statistics across cubes and
+    assign the resulting output cubes to the output_products.
+
+    This function separates the ESMValCore internals (products) from the actual
+    statistics function that operates on Iris cubes.
+
+    Parameters
+    ----------
+    products: list
+        list of PreprocessorFile's
+    statistics: list
+        list of strings describing the statistics that will be computed
+    output_products: dict
+        dict of PreprocessorFile's with statistic names as keys.
+    span: str
+        'full' or 'overlap', whether to calculate the statistics on the time
+        union ('full') or time intersection ('overlap') of the cubes.
 
-    if output_products:
-        products |= statistic_products
-        return products
-    return statistic_products
+    Returns
+    -------
+    set
+        set of PreprocessorFiles containing the computed statistics cubes.
+    """
+    # Extract cubes from products
+    cubes = [cube for product in products for cube in product.cubes]
+
+    # Compute statistics
+    if len(cubes) < 2:
+        logger.info('Found only 1 cube; no statistics computed for %r',
+                    list(products)[0])
+        statistics_cubes = {statistic: cubes[0] for statistic in statistics}
+    else:
+        statistics_cubes = multicube_statistics(cubes=cubes,
+                                                statistics=statistics,
+                                                span=span)
+
+    # Add cubes to output products and log provenance
+    statistics_products = set()
+    for statistic, cube in statistics_cubes.items():
+        statistics_product = output_products[statistic]
+        statistics_product.cubes = [cube]
+
+        for product in products:
+            statistics_product.wasderivedfrom(product)
+
+        logger.info("Generated %s", statistics_product)
+        statistics_products.add(statistics_product)
+
+    return statistics_products
+
+
+def multi_model_statistics(products: set,
```

Comment:
Note that this function is part of the public API. Changing it so it no longer accepts cubes is a breaking change that will probably break diagnostic code. Would it be possible to make it work again that if you put cubes in, you get cubes out? And of course the docstring needs to be good as @valeriupredoi already pointed out.

Reply:
shucks! I totally missed that - very good point @bouweandela - yes, @Peter9192 please see this and revise the approach :+1:

Reply:
Yes I figured that out as well but I'm struggling to find an appropriate solution.

Reply:
@Peter9192 Didn't we talk about making `multi_model_statistics` cube only (not to mess with the public API) and simply pass the arguments to `multi_cube_statistics`. Then use `_multi_model_statistics` with products for internal use? Or am I confusing 2 PRs now?

Reply:
After a lot of contemplating I have accepted that the public API function will have to accept both products and cubes as possible inputs. I had rather hidden those products from the public altogether, but this becomes very ugly very quickly. So I have now restored the public `multi_model statistics` as the single entry point that returns different functions based on its input type.